### PR TITLE
Querying based on time

### PR DESCRIPTION
**Pull request by tomyyyD**
Range and time queries. A bit large on this PR, but a lot is similar logic that gets the files, so it shouldn't be too hard to look over.

### CODE REVIEW

File: rexdb.py [hunk 1]
```diff
@@ -174,63 +217,233 @@ def write_file(self, bytes_data: bytes) -> None:
...
+        folders = []
+        files = []
+        found_folder = False
+
+        try:
+            with open(self.db_map, "rb") as fd:
+                contents = fd.read()
+                if len(contents) != 0:
+                    lines = int(len(contents) / 12)
+                    for i in range(lines):
```
Comment (thetazero) [inline rexdb.py]: should be able to read 12 bytes as a time to avoid this indexing arithmetic.

File: rexdb.py [hunk 2]
```diff
@@ -174,63 +217,233 @@ def write_file(self, bytes_data: bytes) -> None:
...
+                        folders = [self.folders]
+                    if end_time <= end:
+                        folders.append(self.folders)
+        except Exception as e:
+            print(f"couldn't access db map: {e}")
+
+        found_file = True
+        for folder in folders:
+            try:
+                with open(f"db_{self.db_num}/{folder}/{folder}.map", "rb") as fd:
```
Comment (thetazero) [inline rexdb.py]: Maybe try `f"db_{self.db_num}/{folder}/.map"`

File: rexdb.py [hunk 3]
```diff
@@ -174,63 +217,233 @@ def write_file(self, bytes_data: bytes) -> None:
...
+                    contents = fd.read()
+                    if len(contents) != 0:
+                        lines = int(len(contents) / 12)
+                        for i in range(lines):
+                            (start_time, end_time, num) = struct.unpack("iii", contents[i * 12: i * 12 + 12])
+                            if (start <= start_time and start_time <= end or start <= end_time and end_time <= end):
```
Comment (thetazero) [inline rexdb.py]: add parentheses because I have skill issues and don't remember  order of operations for and and or.

Comment (thetazero) [reply rexdb.py]: Can also do the python thing of
```python
(start_time <= start <= end_time) or (start_time <= end <= end_time)
```

File: rexdb.py [hunk 4]
```diff
@@ -174,63 +217,233 @@ def write_file(self, bytes_data: bytes) -> None:
...
+        found_folder = False
+        found_file = False
+
+        # finding folder from DB_map
+        try:
+            with open(self.db_map, "rb") as fd:
+                contents = fd.read()
+                if len(contents) != 0:
+                    lines = int(len(contents) / 12)
+                    for i in range(lines):
```
Comment (thetazero) [inline rexdb.py]: I mean something like this:
```python3
while (data := f.read(12)) and len(data) == 12:
    print(data)
```

File: rexdb.py [hunk 5]
```diff
@@ -245,5 +458,52 @@ def col(self, i):
...
+        The precision of this function goes only to the nearest second because of the restrictions
+        of struct_time
+        """
+        tfloat = time.mktime(t)
+        filepath = self._file_manager.location_from_time(tfloat)
+        if tfloat < self._init_time:
+            raise ValueError("time is before database init time")
+        try:
+            with open(filepath, "rb") as fd:
+                for _ in range(self._file_manager.lines_per_file):
```
Comment (thetazero) [inline rexdb.py:460]: We can probably make a read file method

Comment (thetazero) [reply rexdb.py:460]: I think 
```python
for _ in range(self._file_manager.lines_per_file):
                    raw_data = fd.read(self._packer.line_size)
                    data = self._packer.unpack(raw_data)
                    if (data[0] == tfloat):
                        return data
```
is duplicated

File: rexdb.py [hunk 6]
```diff
@@ -174,63 +217,233 @@ def write_file(self, bytes_data: bytes) -> None:
...
+        when new folder is created, adds current timestamp to the
+        database map as a start time.
+
+        When a folder is finished being written to this function will
+        add the timestamp as that folder's end time.
+
+        writes a struct of int (file number), float (start time),
+        float (end time) to the map
+        """
+        # print(self.folder_start_time, t)
```
Comment (thetazero) [inline rexdb.py]: prints should be removed

**Review by thetazero — CHANGES_REQUESTED** [review]
This is great, a few small tweaks then I think its ready

File: rexdb.py (hunk 5 above)
Comment (tomyyyD) [reply rexdb.py:460]: the other part that uses similar logic does not immediately return rather it appends to a list, and the comparisons are slightly different so I don't think i can abstract this to a method.

Comment (thetazero) [reply rexdb.py:460]: My bad

**Review by thetazero — APPROVED** [review]
Looks good

**Pull request merged**